src/xrc-tests/src/tests/caching.rs

```diff
@@ -0,0 +1,141 @@
+            class: AssetClass::Cryptocurrency,
+        },
+    };
+
+    let responses = mock_responses::exchanges::build_responses(
+        "ICP".to_string(),
+        timestamp_seconds,
+        |exchange| match exchange {
+            xrc::Exchange::Binance(_) => Some("3.91"),
+            xrc::Exchange::Coinbase(_) => Some("3.92"),
+            xrc::Exchange::KuCoin(_) => Some("3.92"),
+            xrc::Exchange::Okx(_) => Some("3.90"),
+            xrc::Exchange::GateIo(_) => Some("3.90"),
+            xrc::Exchange::Mexc(_) => Some("3.911"),
+            xrc::Exchange::Poloniex(_) => Some("4.005"),
+        },
+    )
+    .chain(mock_responses::exchanges::build_common_responses(
+        request.quote_asset.symbol.clone(),
+        timestamp_seconds,
+    ))
+    .chain(mock_responses::stablecoin::build_responses(
+        timestamp_seconds,
+    ))
+    .chain(mock_responses::forex::build_responses(now))
+    .collect::<Vec<_>>();
+    let container = Container::builder()
+        .name("caching")
+        .exchange_responses(responses)
+        .build();
+
+    run_scenario(container, |container| {
+        let run_scenario_instant = Instant::now();
+        while run_scenario_instant.elapsed().as_secs() <= ONE_MINUTE_SECONDS {
+            let iteration_instant = Instant::now();
+            let result = container
+                .call_canister::<_, GetExchangeRateResult>("get_exchange_rate", &request)
+                .expect("Failed to call canister for rates");
+
+            samples.push(ScenarioResult {
+                result,
+                time_passed_millis: iteration_instant.elapsed().as_millis(),
+            });
+        }
+
+        // Check that all samples were successful.
+        assert!(samples.len() > 1);
+        for sample in &samples {
+            match &sample.result {
+                Ok(exchange_rate) => {
+                    assert_eq!(exchange_rate.base_asset, request.base_asset);
+                    assert_eq!(exchange_rate.quote_asset, request.quote_asset);
+                    assert_eq!(exchange_rate.timestamp, timestamp_seconds);
+                    assert_eq!(exchange_rate.metadata.base_asset_num_queried_sources, 7);
+                    assert_eq!(exchange_rate.metadata.base_asset_num_received_rates, 7);
+                    assert_eq!(exchange_rate.metadata.quote_asset_num_queried_sources, 7);
+                    assert_eq!(exchange_rate.metadata.quote_asset_num_received_rates, 7);
+                    assert_eq!(exchange_rate.metadata.standard_deviation, 3_483_761);
+                    assert_eq!(exchange_rate.rate, 88_587_570);
+                }
+                Err(error) => panic!("Received an error from the XRC: {:?}", error),
+            };
+        }
+
+        // Compare the response times of the samples to ensure the cache mechanism was used
+        // for subsequent requests after the first.
+        //
+        // The subsequent requests should arrive in at least half the time.
```

Comment:
The problem is dependent on the worker setup. On my devenv, it's significantly less, but on a GitHub worker, it can be significantly less than half to around half.

Reply:
Okay, that's unfortunate. Since you said that it can be around half, is there a risk that the test is flaky?